Add CalcLength

nattridgell opened the pull request:
Adds:
- CalcLength constructor
- A couple of tests for CalcLength
- fromValue / fromDictionary in LengthValue calls
  corresponding constructors for SimpleLength / CalcLength
- extremely simple tests for fromValue / fromDictionary
- remove unnecessary tags for consistency
  (@constructor, @enum tags)

Commit:
Add CalcLength

Adds:
 - CalcLength constructor
 - A couple of tests for CalcLength
 - fromValue / fromDictionary in LengthValue calls
   corresponding constructors for SimpleLength / CalcLength
 - extremely simple tests for fromValue / fromDictionary
 - remove unnecessary tags for consistency
   (@constructor, @enum tags)

wilddamon commented on src/calc-length.js:
```diff
@@ -0,0 +1,52 @@
+      if (typeof value == 'number') {
+        this[type] = value;
+      } else if (typeof value == 'string') {
+        var nValue = Number.parseFloat(value);
```
Need to check with Shane if we care about parsing strings here

nattridgell replied:
I'll remove it for now and then we can add it back in later if need be.

wilddamon commented on test/js/length-value.js:
```diff
@@ -0,0 +1,13 @@
+suite('LengthValue', function() {
+  test('fromValue returns instanceof a LengthValue and StyleValue', function() {
```
Should these also check for SimpleLength and CalcLength types here?

Commit:
Modified behaviour of LengthValue.fromDictionary

LengthValue.fromDictionary method now:
 - throws an error if an empty CalcDictionary is given
   (also implemented in CalcLength)
 - if only one valid length in the CalcDictionary,
   creates a SimpleLength
 - if >1 valid length in CalcDictionary,
   creates a CalcLength
 - added / altered tests to suit new behaviour

Commit:
Update logic flow in LengthValue.fromDictionary

Add numLengthsFound variable to control flow.
 - Break the for loop as soon as numLengthsFound > 1
   (a CalcLength must be made)
 - if numLengthsFound is 0: no valid length / type in dictionary
 - etc.
 - Style note: put return statements at end of function.

nattridgell commented:
I hope this is what you meant :-)

wilddamon commented on src/calc-length.js:
```diff
@@ -0,0 +1,51 @@
+
+    var hasValidLength = false;
+
+    // NOTE: Does not check dictionary for invalid fields.
```
Please add a comment saying the check is done in fromDictionary.

wilddamon commented on src/length-value.js:
```diff
@@ -65,11 +64,44 @@
+    if (numLengthsFound < 1) {
+      throw new TypeError('A CalcDictionary must have at least one valid length.');
+    } else if (numLengthsFound < 2) {
+      return new SimpleLength(dictionary[firstValidType], firstValidType);
```
As discussed, don't do this for consistency with calc length constructor - a calc with a single value is ok.

wilddamon commented on src/length-value.js:
```diff
@@ -65,11 +64,44 @@
+    } else if (numLengthsFound < 2) {
+      return new SimpleLength(dictionary[firstValidType], firstValidType);
+    } else {
+      return new CalcLength(dictionary);
```
As discussed, we don't want to do this any more - you can move all of this logic into the constructor, and get rid of the part where you create a SimpleLength.
Sorry about that...

Commit:
Revert "Modified behaviour of LengthValue.fromDictionary"

This reverts commit bdbb56126276ad33bbeeb39dd4c1ba3dae379e8f.

Commit:
Commenting error

Fix typo mistakes.

wilddamon commented:
LGTM
